pkg/controller/history/controller_history.go

```diff
@@ -125,15 +126,19 @@ func EqualRevision(lhs *apps.ControllerRevision, rhs *apps.ControllerRevision) b
 		return lhs == rhs
 	}
 	if hs, found := lhs.Labels[ControllerRevisionHashLabel]; found {
-		hash, err := strconv.ParseInt(hs, 10, 32)
-		if err == nil {
+		hash, err := strconv.ParseUint(hs, 10, 32)
```

Comment:
I agree that this should be changed to `ParseUint`, but I'm not sure that's the root problem. I believe the intention of EqualRevision is to deep-compare the revision data; the hash should be conceptually irrelevant. My guess is that we compare the hashes as an optimization to avoid doing the deep-compare when possible.

If that's true, I think there is a bug in this code below:

```go
if lhsHash != nil && rhsHash != nil && *lhsHash != *rhsHash {
  return false
}
```

This code implies that the hashes *must* match in order for two revisions to be equal. Instead, I would expect this (note `!=` => `==` and `false` => `true`):

```go
if lhsHash != nil && rhsHash != nil && *lhsHash == *rhsHash {
  return true
}
```

This implies that if the hashes are valid and they match, we can assume the contents are equal and skip the deep-compare. If the hashes are missing/invalid or don't match (which may be due to collisionCount changing), then we should ignore the hashes and do the actual deep-compare.

Reply:
I got a quick update from @kow3ns who pointed out that we can't trust the hashes being equal, since the hash algorithm we use has a relatively high probability of collisions for different data.

That means the code I said I expected above would also be wrong. Given that, I'm wondering if we should just do the deep-compare every time in this function and ignore the hashes. We'll see what the others think who are more familiar with the collision avoidance feature.

Reply:
Going to try to restate to make sure I have the right understanding

Suppose the intention of EqualRevision should be to return true if and only if the revisions "data" field are deep equal. 
Then `EqualRevision` should ignore the hashes entirely since
* hashes being equal can mean either 
    1. "data" fields were unequal, a (common) collision caused the hashes to be equal
    1. "data" fields were equal, this caused the hashes to be equal
* hashes being unequal can mean either
    1. "data" fields were unequal, this caused the hash to be unequal
    1. "data" fields were equal, the name collision count incrementing caused the hashes to be unequal
Therefore in both cases hash equality doesn't yield definitive signal.

Reply:
> My guess is that we compare the hashes as an optimization to avoid doing the deep-compare when possible.

We shouldn't compare hashes in general. Hash values can change whenever the API schema changes (e.g. add a new field, which would happen in almost every k8s release), and two different data may generate the same hash value and that's why we need hash collision avoidance mechanism.